### PR TITLE
Switch to google-music.js library

### LINKED ISSUE #252: Break out JS integration into separate library
A while ago, I saw this and started writing my own `node-webkit` wrapper to support all platforms. I used parts of the JS files to get something quick and dirty working.

https://github.com/twolfson/google-music-webkit/blob/1.0.0/lib/google-music.js

In the long term, we should have a common repository that we pull from. Ideally, it would support all modern browsers and be an interface that attaches to the Google Music website:

```js
var googleMusic = new GoogleMusic(window);
```

Let me know if you are interested, otherwise I might fork this repo and strip it down to the JS lib with less singleton assumptions.

### PR DESCRIPTION
**Pull request by jacobwgillespie**
This is an early PR that begins the process of removing the custom JavaScript code used to control and monitor the Google Music player and replacing it with [google-music.js](https://github.com/twolfson/google-music.js).  That library was forked from the library inside Radiant and is currently in use in several places, including `google-music-electron`.

In my opinion, we should use a central library for this kind of thing so that we all benefit from shared development resources (especially when Google updates something on their end that breaks us).

This PR currently does not work - I'm still debugging.  I'm opening this early though so that we can get some extra eyes reviewing.

@radiant-player/radiant-player-mac 

Closes #252

### CODE REVIEW

**Comment by twolfson** [comment]
Wow, awesome/great. I think the recent transition from `sj-button` to `paper-button` might have broken some selectors (tried to patch them as best I could but I was in a time crunch) but I'm pretty confident the test suite works for everything `google-music-electron` uses.

**Comment by jadengore** [comment]
@twolfson Thanks for making a library like this! It should be helpful to our project.

**Comment by jacobwgillespie** [comment]
Submitted an issue at twolfson/google-music.js#12 for the error I'm seeing.

Separately, since we're updating element names for both CSS and JS, I wonder if the two combined somehow?  Like having JS style the elements with our themes or having a common element definition file used to generate the CSS and JS.

**Comment by twolfson** [comment]
We could probably add selectors for elements to be themed in `google-music.js` but I think `google-music.js` should more be focused on one thing well (i.e. only handling bindings).

An alterative would be to write a separate library (e.g. `google-music-themer`) to handle the case of theming but that hasn't been in the scope of `google-music-electron`.

**Comment by chrismou** [comment]
@jacobwgillespie Is there a blocker on this at the minute? Was thinking of jumping on to help things along a bit this weekend, but don't want to step on your toes if you've already got it in hand

**Comment by MarshallOfSound** [comment]
Looking through this it appears that you are committing gmusic.js to version control.  Technically this means you are versioning an already versioned utility??

Not sure if it is possible with the mac build tooling but wouldn't it make more sense to add that directory downloaded by bower to the `.gitignore` and add a `bower install` as a pre build task or something?  Then you aren't duplicating files accross multiple version control repos

**Comment by jacobwgillespie** [comment]
@chrismou - feel free to investigate if you'd like!  I'll be on a business trip for the next week so not sure if I'll have time to take a look.  The remaining todos are switch to gmusic-utils.js and then figure out why it isn't updating the playing information in the miniplayer.

@MarshallOfSound - yes, totally.  I'd like to actually switch this to npm (since bower is disappearing it seems) and gitignore the npm install step.  When I started this PR, I was trying to avoid an additional dependency - as a parallel, check out #399 where we debated gitignoring the CocoaPods deps and were considering the extra barrier to entry.

**Comment by MarshallOfSound** [comment]
I recently made the move on GPMDP to use node as part of the build process.  To draw a similarity from this to that.

Nuget = CocoaPods
Node = Bower

To me any dependency that you pull in externally should not be in version control, it should be referenced as a dependency somehow, preferably with an exact version / tag to stop upgrade errors.

In my case both node_modules and Nuget directories are ignored.  This stackoverflow question seems relevant 

http://stackoverflow.com/questions/9446644/what-goes-into-your-gitignore-if-youre-using-cocoapods

TLDR: An `npm install` isn't a barrier to using a node module, so `pod install` shouldn't be either

**Comment by jacobwgillespie** [comment]
I agree with you and would love to remove the dependencies personally.

As a counterpoint: https://guides.cocoapods.org/using/using-cocoapods.html#should-i-check-the-pods-directory-into-source-control

**Comment by MarshallOfSound** [comment]
The points for committing to source control seem to be more related to floors in the actual cocoapods methodology (a reference to a github repo rather than an actual package manager).

Tbh it's up to you guys as controllers of the repo to decide what should / shouldn't be here :100: 

I'd say put it to a vote amongst the contributors :cactus:

**Comment by twolfson** [comment]
To prevent solution fixation (e.g. where only 2 ideas are debated), there are other options:

- Use `git submodules`, not great but it does require 1 less tool (e.g. `npm`)
- Use `curl` or `wget` to download tagged JS from raw source
    - `wget https://raw.githubusercontent.com/gmusic-utils/gmusic.js/4.0.3/dist/gmusic.min.js`
    - This could be placed in a `bootstrap.sh` which installs dependencies

**Comment by chrismou** [comment]
I always assumed gmusic.js wouldn't be under source control in the long run. However you do it is OK with me - easier if it's locked under suede control, but I'm completely fine using npm. But as with #399, it does create another barrier to entry and I know people that have struggled getting node setup properly (not full time devs, admittedly). The wget idea could work, but it does feel a bit ropey ("Step 1: clone repo Step 2: run this random shell script)

With cocoapods - someone did say it could sometimes be a bit of a hassle. I'm happy to move to pulling in dependencies ourselves (this is pretty much the defacto standard in php) but less so if it puts people off contributing

**Comment by chrismou** [comment]
Woops, hit the wrong button :-/

**Comment by MarshallOfSound** [comment]
@chrismou  Getting node up and running on mac is as simple as 

```bash
ruby -e "$(curl -fsSL https://raw.githubusercontent.com/Homebrew/install/master/install)"
brew install node
```

isn't it?  Long time since I've had to install node from scratch :smile_cat:

**Comment by chrismou** [comment]
Not sure, I tend to use nvm! I'm not even sure why struggled so much, but tbf to them they were graphic designers, not devs :-P

**Comment by chrismou** [comment]
Mini player and associated bits are working again.

2 main issues left now (that I've spotted) are:

- Scrobbling - I think the duration format is different in gmusic.js, so currently it's not getting past the percentage played check
- Ratings - Click to thumb up a song will mark it as liked on Google Music, but none of the backend stuff happens - updating icons, submitting to last.fm (if enabled), etc.  I suspect it may have something to do with how it worked before (https://github.com/radiant-player/radiant-player-mac/blob/master/radiant-player-mac/js/main.js#L170) but I haven't had time to look into it yet

I'm actively looking into the scrobbling issue - feel free to jump on the rating issue if you fancy it. Just let me know here so we both don't end up fixing the same thing :-)

I'll merge in master in a bit. That should be fun :+1:

**Comment by chrismou** [comment]
Good news - merge from master (including your star rating removal commit) seems to have fixed ratings.

Will try and get scrobbling nailed tonight if I get time

The irony is now I'm spending time fixing something that I'm going to start work on replacing as soon as gmusic.js has been integrated ;-)

**Comment by chrismou** [comment]
Calculated scrobbled time just had to be divided by 1000. I guess we previously used seconds and gmusic uses microseconds?

Anyways, that's now sorted too.  I'll tidy up all my debug and push everything back tonight (like, actual tonight, not that fake tonight from earlier than turned out to be about 15 mins later :wink:)

I guess the next step for us to is dogfood it for a couple of days and make sure nothing subtle has broken in the background. I'm in the office for the rest of the week so I'll stick to using this branch.

**Comment by jacobwgillespie** [comment]
gmusic probably uses milliseconds...

I'll run off this branch for a bit and see if I spot any issues.  :+1:

**Comment by twolfson** [comment]
We altered it in `gmusic.js@2.0.4`, I think it was for consistency that MPRIS expected (Linux desktop music player integration):

https://github.com/gmusic-utils/gmusic.js/blob/4.0.3/CHANGELOG.md

**Comment by chrismou** [comment]
I had an feeling when I threw it to console and saw a number exactly x1000 more than I expected it to be ;-)

@jacobwgillespie Just tidying some code up. I'll rebase all my commits into one and push it back. If I come across any other issues I'll post here before I do anything in case you're already on it

**Comment by chrismou** [comment]
Issue: miniplayer shows 2 thumbs up graphics. Bottom one seems to be the actual thumb up, tops is thumbs down. Clicking either works as expected

**Comment by chrismou** [comment]
Some weirdness going on here. I'm dumping mode out to JS console for change:shuffle and change:repeat. Just clicking play on a song, I'm getting 10 outputs for each.  6 shuffle, then 6 repeat, 2 shuffle, 2 repeat, 2 shuffle, 2 repeat. mode returns undefined for each.

Clicking the repeat button manually outputs 6 repeat, 2 shuffle, then 4 repeat. Again, all undefined.

@twolfson Is this normal?  Code is here: https://github.com/radiant-player/radiant-player-mac/blob/switch-to-google-music-js/radiant-player-mac/js/main.js#L46  -- pretty vanilla implementation.  Had a quick glance at the gmusic and the selectors seem to match up (from what I can see).

**Comment by MarshallOfSound** [comment]
Without looking too deep into your implementation it sounds like you're adding event handlers multiple times.  Possibly your code is executing more than once.  Shove a console log before adding the event handlers and see how many times it is called

**Comment by chrismou** [comment]
@MarshallOfSound Handlers are only being set once, judging by the console output. Plus I'm not sure that would explain it returning undefined.

I'll have to dig into this a bit more later when I'm not at the office

**Comment by twolfson** [comment]
The repeat/shuffle implementations leverage `MutationObserver` which look and fire on any relevant DOM mutations. It looks like we don't do anything like verifying the previous value changed so we are prob firing due to multiple attributes changing:

https://github.com/gmusic-utils/gmusic.js/blob/4.0.3/lib/main.js#L366-L386

whereas in `addObserver` we make sure the value changed:

https://github.com/gmusic-utils/gmusic.js/blob/4.0.3/lib/main.js#L348

I'm not sure why the values are `undefined` but we should probably move the `target.value` resolvers to use their `playback.getRepeat`/`playback.getShuffle` equivalents -- no need to reimplement/maintain the same code twice.

In any event, we should prob move this chunk of the thread to `gmusic.js`

**Comment by jacobwgillespie** [comment]
Replaced this with #520.
